modules/examples/sources/threaded-random-generator/threaded-random-generator.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 László Várady <laszlo.varady@balabit.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "threaded-random-generator.h"
+#include "logthrsource/logthrsourcedrv.h"
+#include "atomic.h"
+#include "messages.h"
+#include "str-format.h"
+
+#include <strings.h>
+#include <sys/random.h>
+#include <unistd.h>
+
+struct ThreadedRandomGeneratorSourceDriver
+{
+  LogThreadedSourceDriver super;
+  GAtomicCounter exit_requested;
+
+  guint freq;
+  guint bytes;
+  gboolean urandom;
+};
+
+static gboolean
+_generate_random_bytes(guint8 *random, guint size, gboolean urandom)
+{
+  guint length = 0;
+
+  while (length < size)
+    {
+      gssize rc = getrandom(random + length, size - length, urandom ? 0 : GRND_RANDOM);
+
+      if (rc < 0)
+        {
+          msg_error("Could not generate random bytes", evt_tag_error("error"));
+          return FALSE;
+        }
+
+      length += rc;
+    }
+
+  return TRUE;
+}
+
+
+/* runs in a dedicated thread */
+static void
+_run(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  guint8 *random_bytes = g_malloc(self->bytes);
+
+  const gsize random_hex_str_size = self->bytes * 2 + 1;
+  gchar *random_hex_str = g_malloc(random_hex_str_size);
+
+  while (!g_atomic_counter_get(&self->exit_requested))
+    {
+      if (_generate_random_bytes(random_bytes, self->bytes, self->urandom))
+        {
+          format_hex_string(random_bytes, self->bytes, random_hex_str, random_hex_str_size);
+
+          LogMessage *msg = log_msg_new_empty();
+          log_msg_set_value(msg, LM_V_MESSAGE, random_hex_str, -1);
+
+          log_threaded_source_blocking_post(s, msg);
+        }
+
+      usleep(self->freq * 1000);
+    }
+
+  g_free(random_hex_str);
+  g_free(random_bytes);
+}
+
+static void
+_request_exit(LogThreadedSourceDriver *s)
+{
+  ThreadedRandomGeneratorSourceDriver *self = (ThreadedRandomGeneratorSourceDriver *) s;
+
+  g_atomic_counter_set(&self->exit_requested, TRUE);
+}
```

Comment:
and if we had the should_exit() call, we wouldn't need this method at all.

Reply:
Please see my comment above.